src/App.js

```diff
@@ -2,355 +2,44 @@ import React from "react";
+      var nextResult = updateDisplay(result, button.key, lastEquation);
 
-
-  };
-
-
-
-  calculate = () => {
-    const { result, operate, originalLastNum } = this.state;
-    let finalResult = 0;
-    var tempResult = result;
-
-    // Special case with expression ending with operator, after CE use
-    if ("-+*/".includes(this.getLastChar(result))) {
-      tempResult = result.slice(0, -1);
+      this.setState({
+        result: nextResult,
+        lastEquation: "",
+      });
     }
 
+    if (button.key === "=") {
+      var [nextEquation, finalResult] = runEquation(result, lastEquation);
 
-    tempResult = this.closeParens(tempResult) + "";
-    tempResult = this.addMultiplier(tempResult) + "";
-
-
-    // Handling double equal
-    // If I only have sign + digit(s) + dot + digit(s)
-    const regex = /-{0,1}[0123456789]*(\.[0123456789]*){0,1}/g;
-    const matches = tempResult.match(regex) || [];
-    if (matches.length > 1 && matches[0] === tempResult) {
-      //I need to handle the double equal
-      if ("+-/*".includes(operate) && !isNaN(originalLastNum)) {
-        // finalResult = evaluate(result + operate + originalLastNum) + "";
-        try {
-          finalResult = this.roundedResult(tempResult + operate + originalLastNum) + "";
-        } catch(error) {
-
-          //throw error
-        }
-      }
-    } else {
-      var tempResultString = tempResult;
-      //We need to detect the non nested parenthesis
-      //Open ( and no other ( before the next close
-      // '(' 0123456789-+*/.(not (, not ) ) ')'
-      // ( + any number of the 15 chars + ) : (7+2) (-3.25478/+6.587)
-      const reg = /\(([0123456789/*-+.]*)\)/g;
-      var parenthesisToCalculate = tempResultString.match(reg) || [];
-      while (parenthesisToCalculate.length > 0) {
-        for (var i = 0; i < parenthesisToCalculate.length; i++) {
-          //we extract the first expression ex:  (2+5)
-          let expression = parenthesisToCalculate[i];
-          //We calculate the value ex: 7
-
-          try {
-            let tempResult = round(evaluate(expression), 13);
-
-            //We need to replace the expression by the calculation
-            tempResultString = tempResultString.replace(expression, tempResult);
-
-          } catch (error) {
-            //throw error
-          }
-        }
-
-        parenthesisToCalculate = tempResultString.match(reg) || [];
-      }
-      //In tempResultString we have the last expression withtout any ()
-
-      try {
-       finalResult = round(evaluate(tempResultString), 13);
-      } catch (error) {
-        //throw error
-      }
       this.setState({
-        done: true,
         result: finalResult + "",
+        lastEquation: nextEquation,
+      });
+    } else if (button.key === "AC") {
+      this.setState({
+        result: "",
+        lastEquation: "",
+      });
+    } else if (button.key === "CE") {
+      this.setState({
+        result: result.slice(0, -1),
       });
     }
```

Comment:
@grimxreaper Can we refactor this so there is a single setState at the end?

Reply:
It isn't super clear from all the code referenced, do you mean refactoring to one single setState in the onClick?

Reply:
You got it! Yeah I was essentially highlighting the internals of onClick.

Reply:
So I feel like this should work but I get back loads of red text that say "newTempResult.match is not a function" but it doesn't give me the actual error. Could you give me a hint as to what I would need to fix? 

<img width="602" alt="Screen Shot 2021-07-11 at 1 27 14 PM" src="https://user-images.githubusercontent.com/64915527/125204578-fe314b00-e24b-11eb-8d8c-884e1f735f29.png">

<img width="543" alt="Screen Shot 2021-07-11 at 1 27 24 PM" src="https://user-images.githubusercontent.com/64915527/125204587-01c4d200-e24c-11eb-81c4-1fcdd7d02389.png">

Reply:
I want to see a commit where you successfully reduce it to a single setState. You are making it harder by extracting a function at the same time. This is why small steps are important. Don't skip steps!

Do you know where the error is occurring? Have you tried using console logs, debugger, or commenting out code?

Reply:
Imagine a bug occurred that you didn't catch, then someone has to figure out what about your commit broke things. If it's in separate commits, it's easier to identify with git bisect.

Reply:
How would I reduce it to a single setState without extracting a function? It seems that it is extracting the function that would make a single setState possible...

Reply:
Nope. Totally possible without extracting. In fact, necessary in order to be able to extract it.

Reply:
The idea is you want a function that has everything currently in the onClick except for the react state stuff. In order to extract it, you must first "move" the setState's out of the way. Only then can you extract the code safely. In fact, by doing so you may avoid the bug you're seeing now.

Reply:
Do you mean something like this?
<img width="569" alt="Screen Shot 2021-07-11 at 2 30 49 PM" src="https://user-images.githubusercontent.com/64915527/125206319-a0edc780-e254-11eb-9dd8-1e8fad964820.png">

Reply:
That's better. Any guesses why you might have issues with that?

Reply:
Do you notice the squiggly line?

Reply:
What if you tried only removing one setState and seeing if things still work? Then removing another. Instead of removing them all and seeing what happens.

Reply:
> Do you notice the squiggly line?

Yes, but I'm not sure what to do about that since the variable name will need to be the same for me to only setState once, or that's my understanding at least.

Reply:
> What if you tried only removing one setState and seeing if things still work? Then removing another. Instead of removing them all and seeing what happens.

That's a good idea, I tried it, and it doesn't actually seem to be updating state, even when I go one at a time.